src/prdc/iterator/ExtGenFilmBase.h

```diff
@@ -0,0 +1,255 @@
+
+      /**
+      * Set the value of a specialized sweep parameter.
+      * 
+      * In this class, the two specialized sweep parameters are chi_top
+      * and chi_bottom.
+      * 
+      * \param name  name of the specialized parameter
+      * \param ids  array of integer indices specifying the value to set
+      * \param value  the value to which the parameter is set
+      */
+      void setParameter(std::string name, DArray<int> ids, 
+                                          double value, bool& success);
+
+      /**
+      * Get the value of a specialized sweep parameter.
+      * 
+      * In this class, the two specialized sweep parameters are chi_top
+      * and chi_bottom.
+      * 
+      * \param name  name of the specialized parameter
+      * \param ids  array of integer indices specifying the value to set
+      */
+      double getParameter(std::string name, DArray<int> ids, bool& success) 
+      const;
+
+      /**
+      * Get const chiBottom matrix by reference
+      */
+      DArray<double> const & chiBottom() const;
+
+      /**
+      * Get const chiTop array by reference
+      */
+      DArray<double> const & chiTop() const;
+
+      /**
+      * Get the chi parameter between the bottom wall and species s
+      * 
+      * \param s  species index, 0 <= id < nVertex
+      */
+      double chiBottom(int s) const;
+
+      /**
+      * Get the chi parameter between the top wall and species s
+      * 
+      * \param s  species index, 0 <= id < nVertex
+      */
+      double chiTop(int s) const;
+
+      /**
+      * Get value of normalVecId_ stored in this object
+      */
+      int normalVecId() const;
+
+      /**
+      * Check whether the fields have been generated.
+      */
+      bool isGenerated() const = 0;
+
+      using ParameterModifier::setParameter; // overloaded method
+      using ParameterModifier::getParameter; // overloaded method
+
+   protected:
+
+      /**
+      * Allocate container necessary to generate and store fields.
+      */ 
+      virtual void allocate() = 0;
+
+      /**
+      * Generate the fields and store where the Iterator can access.
+      */
+      virtual void generate() = 0;
+
+      /**
+      * Get the space group name for this system.
+      */
+      virtual std::string getSpaceGroup() const = 0;
+
+      /**
+      * Get the lattice parameters for this system.
+      */
+      virtual FSArray<double, 6> getLatticeParameters() const = 0;
+
+      /**
+      * Get the number of monomer species for this system.
+      */
+      virtual int getNMonomer() const = 0;
+
+      /**
+      * Use mask to determine normalVecId and store value in this object.
+      */
+      virtual void getNormalVecId() = 0;
+
+      /// Lattice parameters associated with the current external fields
+      FSArray<double, 6> parameters_;
+
+      /// chiBottom array associated with the current fields
+      DArray<double> chiBottomCurrent_;
```

Comment:
The distinction between the "current" values of, e.g., chiBottom and chiTop (which are protected) and the variables chiBottom and chiTop are confusing to me.  Why not just have one set of values that might be modified during a sweep?   There may be nothing wrong with this but, as your first code reader, I can't immediately see or guess why you would make a distinction like this.  Consider getting rid of the distinction (if possible) or documenting why it exists.